CGEN error with map having optional sumtype

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.9 5c04f32, press to see full `v doctor` output</summary>

|V full version      |V 0.4.9 45fd7eb.5c04f32
|:-------------------|:-------------------
|OS                  |windows, Microsoft Windows 11 Pro 26100 64-bit
|Processor           |24 cpus, 64bit, little endian, AMD Ryzen 9 5900X 12-Core Processor  
|Memory              |72.43GB/127.92GB
|                    |
|V executable        |S:\repo\vlang\v.exe
|V last modified time|2025-02-03 15:51:54
|                    |
|V home dir          |OK, value: S:\repo\vlang
|VMODULES            |OK, value: C:\Users\john3\.vmodules
|VTMP                |OK, value: S:\temp\v_0
|Current working dir |OK, value: S:\vProjects\vSelenium
|                    |
|Git version         |git version 2.45.1.windows.1
|V git status        |weekly.2025.04-90-g5c04f32f
|.git/config present |true
|                    |
|cc version          |N/A
|gcc version         |gcc (MinGW-W64 x86_64-ucrt-posix-seh, built by Brecht Sanders, r2) 14.2.0
|clang version       |(built by Brecht Sanders, r2) clang version 19.1.1
|msvc version        |N/A
|tcc version         |tcc version 0.9.27 (x86_64 Windows)
|tcc git status      |thirdparty-windows-amd64 b425ac82
|emcc version        |N/A
|glibc version       |N/A

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg -cg main.v && S:\vProjects\vSelenium\main.exe`
See attached file `main.v`

**What did you see?**
```
webdriver/remote/connection_manager.v:3:8: warning: module 'pool (sync.pool)' is imported but never used
    1 | module remote
    2 | 
    3 | import sync.pool {PoolProcessor, PoolProcessorConfig, new_pool_processor}
      |        ~~~~~~~~~
    4 | import net.http
    5 |
webdriver/remote/remote_connection.v:8:8: warning: module 'pool (sync.pool)' is imported but never used
    6 | import encoding.base64
    7 | import maps {merge, merge_in_place}
    8 | import sync.pool
      |        ~~~~~~~~~
    9 | 
   10 | type Any = ?int | ?string
webdriver/common/selenium_manager.v:140:3: warning: unused variable: `arch`
  138 |         s.machine = un.machine
  139 |         s.user_os = os.user_os()
  140 |         arch := if s.user_os in ['linux', 'freebsd', 'openbsd'] { s.machine } else { 'any' }
      |         ~~~~
  141 |         if s.user_os in ['freebsd', 'openbsd'] {
  142 |             log.warn('Selenium Manager binary may not be compatible with ${s.user_os}; verify settings')
webdriver/remote/remote_connection.v:67:2: warning: unused variable: `path_string`
   65 |     }
   66 | 
   67 |     path_string := command_info[1]
      |     ~~~~~~~~~~~
   68 |     
   69 | }
webdriver/chromium/web_driver.v:39:2: warning: unused variable: `executor`
   37 |     drv.service.start()
   38 | 
   39 |     executor := ChromiumRemoteConnection.init(
      |     ~~~~~~~~
   40 |         drv.service.service_url, browser_name, vendor_prefix, kalive, drv.options.ignore_local_proxy, none)
   41 |
S:/temp/v_0/main.01JK6SPDP1YMPET0YXV6N26HGE.tmp.c:846: warning: WINVER redefined

S:/temp/v_0/main.01JK6SPDP1YMPET0YXV6N26HGE.tmp.c:1248: warning: InterlockedIncrement64 redefined

In file included from S:/temp/v_0/main.01JK6SPDP1YMPET0YXV6N26HGE.tmp.c:1628:

s:/repo/vlang/thirdparty/tcc/include/winapi/synchapi.h:13: warning: CreateEvent redefined

In file included from S:/temp/v_0/main.01JK6SPDP1YMPET0YXV6N26HGE.tmp.c:1628:

s:/repo/vlang/thirdparty/tcc/include/winapi/synchapi.h:117: warning: OpenEvent redefined

In file included from S:/temp/v_0/main.01JK6SPDP1YMPET0YXV6N26HGE.tmp.c:1628:

s:/repo/vlang/thirdparty/tcc/include/winap
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

```V
   40 |
main.01JK78A4RYHEH8QSB91F8JYPHN.tmp.c
S:\temp\v_0\main.01JK78A4RYHEH8QSB91F8JYPHN.tmp.c(34260): error C2440: 'initializing': cannot convert from '_option_int' to 'byte'
S:\temp\v_0\main.01JK78A4RYHEH8QSB91F8JYPHN.tmp.c(34266): error C2440: 'initializing': cannot convert from 'string' to '_option_int *'
S:\temp\v_0\main.01JK78A4RYHEH8QSB91F8JYPHN.tmp.c(34270): error C2440: 'initializing': cannot convert from 'string' to '_option_int *'
S:\temp\v_0\main.01JK78A4RYHEH8QSB91F8JYPHN.tmp.c(34273): error C2440: 'initializing': cannot convert from 'string' to '_option_int *'
```

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -138,7 +138,7 @@ fn (mut g Gen) expr_with_opt(expr ast.Expr, expr_typ ast.Type, ret_typ ast.Type)
 	defer {
 		g.inside_opt_or_res = old_inside_opt_or_res
 	}
-	if expr_typ.has_flag(.option) && ret_typ.has_flag(.option)
+	if expr_typ.has_flag(.option) && ret_typ.has_flag(.option) && !g.is_arraymap_set
 		&& expr in [ast.SelectorExpr, ast.DumpExpr, ast.Ident, ast.ComptimeSelector, ast.AsCast, ast.CallExpr, ast.MatchExpr, ast.IfExpr, ast.IndexExpr, ast.UnsafeExpr, ast.CastExpr] {
 		if expr in [ast.Ident, ast.CastExpr] {
 			if expr_typ.idx() != ret_typ.idx() {

--- a/vlib/v/tests/options/option_map_set_test.v
+++ b/vlib/v/tests/options/option_map_set_test.v
@@ -1,0 +1,12 @@
+type Any = ?int | ?string
+
+struct TestConfig {
+	timeout ?int
+}
+
+fn test_main() {
+	mut r := TestConfig{}
+	mut m := map[string]?Any{}
+	m['timeout'] = r.timeout
+	assert m.str() == "{'timeout': Option(Any(Option(none)))}"
+}

